admin/console/dangerzone-for-experts.rst

```diff
@@ -29,76 +29,85 @@ Deleting Customers
 
    As such, the examples below will delete not only the specified customers, but **all tickets associated with them**, as well.
 
-.. code-block:: ruby
+Step 1: Select customers by email address
+   .. code-block:: ruby
+
+      >> customers = User.where(email: %w[customer@example.com customer@example.org])
+
+      >> customers = customers.joins(roles: :permissions)
+                              .where(roles: { active: true })
+                              .where(permissions: { name: 'ticket.customer', active: true })
+                              .where.not(id: 1)
 
-   # Select customers by email address
-   >> customers = User.where(email: %w[customer@example.com customer@example.org])
+Step 2: Preview affected users & tickets
+   .. code-block:: ruby
 
-   >> customers = customers.joins(roles: :permissions)
-                           .where(roles: { active: true })
-                           .where(permissions: { name: 'ticket.customer', active: true })
-                           .where.not(id: 1)
+      >> puts customers.map do |user|
+           "Customer #{user.fullname}/#{user.id} has #{Ticket.where(customer_id: user.id).count} tickets #{Ticket.where(customer_id: user.id).pluck(:number)}"
+         end.join("\n")
```

Comment:
`.join` seems not the be available within a default Zammad rails console and thus does not work:

```
#<Enumerator:0x000000000b2e1020>
Traceback (most recent call last):
        1: from (irb):19
NoMethodError (undefined method `join' for nil:NilClass)
```

I tried to solve this, however, the console returns these kind of "strings":
```
#<User:0x000000000bbecd40>
#<User:0x000000000bbecc00>
```

This does not return any users / tickets as I expected when running this code (to solve the join issue):
```
2.5.5 :020 > puts customers.map do |user|
2.5.5 :021 >                "Customer #{user.fullname}/#{user.id} has #{Ticket.where(customer_id: user.id).count} tickets #{Ticket.where(customer_id: user.id).pluck(:number)} \n"
2.5.5 :022?>   end
```

the result of the above is this:
```
#<Enumerator:0x000000000bbe3e48>
 => nil
```

Reply:
I fiddled a bit, this code does work:

```
puts customers.map {|user|
           "Customer #{user.fullname}/#{user.id} has #{Ticket.where(customer_id: user.id).count} tickets #{Ticket.where(customer_id: user.id).pluck(:number)}"
         }.join("\n")
```

Reply:
Aha! There's a couple of funny things going on here; let me see if I can elucidate them for you.

1. > `.join` seems not the be available within a default Zammad rails console

   The `#join` method is definitely available, but **we're calling it on the wrong thing**. The traceback/error message contains a crucial hint: "undefined method \`join' for nil:NilClass":

   ```ruby
   # see that #join is defined on arrays
   >> [1, 2, 3].join("\n")
   => "1\n2\n3"

   # but not on nil
   >> nil.join("\n")
   NoMethodError: undefined method `join' for nil:NilClass
   ```

   What this means is that everything leading up to the `.join` in the third line is evaluating to `nil` when it should be evaluating to an array.

2. So what is evaluating to `nil`, and why does switching from `do`/`end` to `{`/`}` fix it??

   ```ruby
   # our code sample, simplified
   puts customers.map { block }.join("\n")

   # what we mean
   puts(customers.map { block }.join("\n"))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ intended console output

   # what the interpreter sees
   (puts(customers.map) { block }).join("\n")
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluates to nil
   ```

I will address this in my next commit, as well. Great looking out. Truly, truly grateful for your sharp eyes.

Reply:
I opted for a combination of your `{`/`}` solution and a heredoc, so that the entire `puts` invocation could be kept on one single line.